Add runner for instruction count benchmarks.

robieta opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#54652 Add runner for instruction count benchmarks.**
* #54651 extend benchmarks

This PR adds a fairly robust runner for the instruction count microbenchmarks. Key features are:

* Timeout and retry. (In rare cases, Callgrind will hang under heavy load.)
* Robust error handling and keyboard interrupt support.
* Benchmarks are pinned to cores. (Wall times still won't be great, but it's something.)
* Progress printouts, including a rough ETA.

Differential Revision: [D27537823](https://our.internmc.facebook.com/intern/diff/D27537823)

Commit:
Add runner for instruction count benchmarks.

[ghstack-poisoned]

xuzhao9 commented on benchmarks/instruction_counts/execution/runner.py:
```diff
@@ -0,0 +1,240 @@
+        super().__init__()
+
+
+class CorePool:
```
In TorchBench CI, we use the environment variable `GOMP_CPU_AFFINITY` and the `taskset` tool to set cpu affinity of the benchmark. It looks like the `CorePool` class sets cpu affinity similar to `taskset`, so I guess running the instruction count benchmark only requires setting `GOMP_CPU_AFFINITY` and `core_pool`?

robieta replied:
I think you have put much more thought into stabilizing the environment than I have. I added `GOMP_CPU_AFFINITY` to the launch command; thanks for the tip!

xuzhao9 commented on benchmarks/instruction_counts/execution/work.py:
```diff
@@ -0,0 +1,204 @@
+
+    @property
+    def cmd(self) -> str:
+        cmd: List[str] = []
```
Similar question as above, maybe we should return both command and environment here, so that we can set the "GOMP_CPU_AFFINITY" environment variable as well?

robieta replied:
Indeed.

bitfort commented on benchmarks/instruction_counts/execution/runner.py:
```diff
@@ -0,0 +1,240 @@
+import math
```
Please give the 1 line description of the file at the top for skimmers and lazy people like me :)

bitfort commented on benchmarks/instruction_counts/execution/runner.py:
```diff
@@ -0,0 +1,240 @@
+
+class CorePool:
+    def __init__(self, min_core_id: int, max_core_id: int) -> None:
+        assert min_core_id >= 0
```
Love the handling of assertions and types here! Great!!

robieta replied:
Thanks! One think that MyPy does that's pretty cool is type narrowing. So if you have `x: Optional[int]` and you `assert x is not None`, for the rest of the scope x is treated as `int` rather than `Optional[int]`. (And same for `Union`, etc.) Not directly applicable here, but I bring it up because just making things mypy-strict compliant has made me subconsciously start adding a bunch more asserts.

bitfort commented on benchmarks/instruction_counts/execution/runner.py:
```diff
@@ -0,0 +1,240 @@
+    def reserve(self, n: int) -> Optional[str]:
+        """Simple first-fit policy.
+
+        If successful, return a string for `taskset`. Otherwise, return None.
```
Can you clarify that n is the number of CPUs we want to reserve in a contiguous block?

robieta replied:
Indeed. I added a docstring to `CorePool` explaining some of the reasoning.

bitfort commented on benchmarks/instruction_counts/execution/runner.py:
```diff
@@ -0,0 +1,240 @@
+        super().__init__()
+
+
+class CorePool:
```
So my understanding of this class is that it hands out CPUs (by ID) in blocks which can be returned. It seems like you go out of your way to get contiguous blocks of IDs. I don't see anything specifically wrong with that, though I also don't know if contiguous numbers are particularly important either. What's the thought process here?

robieta replied:
The TL;DR is it's to reduce the chance of spanning two different CPU caches / NUMA nodes unnecessarily. (Added a docstring to that effect as well.)

bitfort commented on benchmarks/instruction_counts/execution/runner.py:
```diff
@@ -0,0 +1,240 @@
+            self._force_shutdown(verbose=True)
+            raise
+
+    def _run(self) -> Dict[WorkOrder, WorkerOutput]:
```
Nice clean structure to this, love it.

robieta replied:
Thanks!

bitfort commented on benchmarks/instruction_counts/execution/work.py:
```diff
@@ -0,0 +1,204 @@
+        with open(self._communication_file, "wb") as f:
+            pickle.dump(self._work_order.timer_args, f)
+
+        self._proc = subprocess.Popen(
```
Consider printing or having a way for this command to be printed... this is exceptionally useful for debugging subprocesses which fail :)

robieta replied:
This is already part of `TimeoutExpired`, and I added it the general `WorkerFailed` as well. Funny you should bring it up though; I literally had a bug when adding `GOMP_CPU_AFFINITY` that caused a malformed command. Once it was printed the bug was obvious. So it's already paid for itself!

bitfort commented on benchmarks/instruction_counts/execution/work.py:
```diff
@@ -0,0 +1,204 @@
+        self._cpu_list = cpu_list
+        self._start_time = time.time()
+        self._communication_file = os.path.join(get_temp_dir(), f"{uuid.uuid4()}.pkl")
+        with open(self._communication_file, "wb") as f:
```
The communication file thing makes me a little queazy. A few thoughts here:
Can you used NamedTemporaryFile? Are we sure we don't need flock (I suppose the life cycle of the subprocess keeps it safe)? Is standard input & standard output not appropriate for passing this data?

robieta replied:
Yeah, in general I'm relying on lifetimes for safety. The process doesn't get launched until the input is written, and the output isn't read until the child has terminated. So I don't think I need flock, though if I'm wrong by all means let me know. 

The reason I don't have to use `NamedTemporaryFile` is that `_communication_file` lives in `get_temp_dir()`, and that dir is guarded by an `atexit.register(shutil.rmtree, ...)`, so rather than worrying after individual files I just guaranteed that hell or high water everything gets cleaned up. (Except something like SIGKILL, of course. But I can't really help that...) I don't like using stdin/stdout because the Python PIPE implementation has as fixed 16k size and is prone to hang if that is exceeded. (Plus I don't want to pull pickle dumps out of a stream...)

That said, this is just my thought process. It is entirely possible that I'm violating some best practices. I'm going to leave this unresolved until you're satisfied.

bitfort commented on benchmarks/instruction_counts/execution/work.py:
```diff
@@ -0,0 +1,204 @@
+        return self._proc.duration
+
+    @property
+    def ready(self) -> bool:
```
I would suggest this is not a property -- properties are idempotent - this has semantics which vary based on when / how many times it is called. Also, the semantics are kind of confusing here because it combines as a check with control logic. What options we have for structuring this?

robieta replied:
I would argue that a property is only idempotent if the underlying structure has `const` semantics (whatever that means in Python...), but I suppose that's a question for the philosophers. I changed the name to `check_finished` and removed the property decorator to signal that it's a more active call. Thoughts?

bitfort reviewed (approved):
Approved modulo comments :)
Thanks Taylor!

Commit:
Update on "Add runner for instruction count benchmarks."


This PR adds a fairly robust runner for the instruction count microbenchmarks. Key features are:

* Timeout and retry. (In rare cases, Callgrind will hang under heavy load.)
* Robust error handling and keyboard interrupt support.
* Benchmarks are pinned to cores. (Wall times still won't be great, but it's something.)
* Progress printouts, including a rough ETA.

[ghstack-poisoned]

Commit:
Update on "Add runner for instruction count benchmarks."


This PR adds a fairly robust runner for the instruction count microbenchmarks. Key features are:

* Timeout and retry. (In rare cases, Callgrind will hang under heavy load.)
* Robust error handling and keyboard interrupt support.
* Benchmarks are pinned to cores. (Wall times still won't be great, but it's something.)
* Progress printouts, including a rough ETA.

[ghstack-poisoned]

Commit:
Update on "Add runner for instruction count benchmarks."


This PR adds a fairly robust runner for the instruction count microbenchmarks. Key features are:

* Timeout and retry. (In rare cases, Callgrind will hang under heavy load.)
* Robust error handling and keyboard interrupt support.
* Benchmarks are pinned to cores. (Wall times still won't be great, but it's something.)
* Progress printouts, including a rough ETA.

Differential Revision: [D27537823](https://our.internmc.facebook.com/intern/diff/D27537823)

[ghstack-poisoned]
